### PR TITLE
planner, sessionctx : Add 'last_plan_from_cache' to help know whether sql's plan is from plan cache

### LINKED ISSUE #16127: we should add extra info to inform that a query plan is from plan cache
## Feature Request

**Is your feature request related to a problem? Please describe:**

prepared-plan-cache.enabled=true;
```
use test;
create table t(a int);
prepare stmt from "select * from t";
select connection_id();
execute stmt;
# explain for connection id;
```
I don't know whether the last sql's plan is from plan cache.

**Describe the feature you'd like:**

mysql> explain for connection 1;
+-----------------------+---------+-----------+---------------+--------------------------------+
| id                    | estRows | task      | access object | operator info                  |
+-----------------------+---------+-----------+---------------+--------------------------------+
| TableReader_5         | 3.00    | root      |               | data:TableFullScan_4, **plan cache: true**          |
| └─TableFullScan_4 | 3.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-----------------------+---------+-----------+---------------+--------------------------------+
2 rows in set (0.01 sec)

### PR DESCRIPTION
**Pull request by lawyerphx**
### What problem does this PR solve?

Issue Number: close #16127   

Problem Summary:

### What is changed and how it works?

What's Changed:
* add a session scope system variable `last_plan_from_cache` to help decide whether the plan cache got hit.

How it Works:
* last_plan_from_cache` indicates whether the last executed statement hit the plan cache. (true for yes, false for no) 
### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 
The default value of `last_plan_from_cache` is false. Set operations of `last_plan_from_cache` will be ignored, and a warning will be given about that.

### CODE REVIEW

**Comment by lawyerphx** [comment]
\run-check-dev

**Comment by lawyerphx** [comment]
\run-unit-tests

**Comment by lawyerphx** [comment]
PTAL~ @francis0407 , @zz-jason

File: planner/core/common_plans.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/sessionctx/variable"
```
Comment (zz-jason) [inline planner/core/common_plans.go:20]: please reorganize the imported packages to two parts:
- packages from golang
- packages from third-part.

File: planner/core/common_plans.go [hunk 2]
```diff
@@ -282,6 +289,10 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
 		} else {
 			planCacheCounter.Inc()
 		}
+		err = e.addHitInfo(sctx, "1")
```
Comment (zz-jason) [inline planner/core/common_plans.go:292]: Please oo not use magic strings in the code. You can pre-define them like this:
```go
const (
	hintOptOn  string = "1"
	hintOptOff string = "0"
)
```

and use it here like this:
```go
		err = e.addHitInfo(sctx, hintOptOn)
```

File: planner/core/common_plans.go [hunk 3]
```diff
@@ -307,12 +318,16 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
 				}
 			}
 			if planValid {
+				err := e.addHitInfo(sctx, "1")
```
Comment (zz-jason) [inline planner/core/common_plans.go:321]: ditto

File: planner/core/common_plans.go [hunk 4]
```diff
@@ -263,6 +264,12 @@ func (e *Execute) checkPreparedPriv(ctx context.Context, sctx sessionctx.Context
 	return err
 }
 
+func (e *Execute) addHitInfo(sctx sessionctx.Context, opt string) error {
```
Comment (zz-jason) [inline planner/core/common_plans.go:267]: The function name is `addHitInfo`, but it actually manipulates the variable `TiDBFoundInPlanCache`. Please rename it to another proper name.

BTW, it's better to change the `opt` to boolean type. Thus we don't need to pass magic string like "1" or "0" to this function.

File: planner/core/prepare_test.go [hunk 5]
```diff
@@ -676,6 +676,39 @@ func (s *testPlanSerialSuite) TestPlanCacheUnionScan(c *C) {
...
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id int)")
+	tk.MustExec("insert into t values (1),(2),(3),(4)")
+	tk.MustExec("prepare stmt from 'select * from t where id=?'")
+	tk.MustExec("set @doma = 1")
+	tk.MustQuery(`show session variables like "last_statement_found_in_plan_cache"`).Check(testkit.Rows(
+		"last_statement_found_in_plan_cache 0",
+	))
+	tk.MustQuery("execute stmt using @doma").Check(testkit.Rows("1"))
+	tk.MustQuery(`show session variables like "last_statement_found_in_plan_cache"`).Check(testkit.Rows(
```
Comment (zz-jason) [inline planner/core/prepare_test.go:707]: ditto

File: sessionctx/variable/session.go [hunk 6]
```diff
@@ -594,6 +594,9 @@ type SessionVars struct {
 	// WindowingUseHighPrecision determines whether to compute window operations without loss of precision.
 	// see https://dev.mysql.com/doc/refman/8.0/en/window-function-optimization.html for more details.
 	WindowingUseHighPrecision bool
+
+	// PlanInCache indicates whether the last statement was found in plan cache
+	PlanInCache bool
```
Comment (zz-jason) [inline sessionctx/variable/session.go:599]: s/PlanInCache/FoundInPlanCache/

File: planner/core/prepare_test.go [hunk 7]
```diff
@@ -676,6 +676,39 @@ func (s *testPlanSerialSuite) TestPlanCacheUnionScan(c *C) {
...
+		"last_statement_found_in_plan_cache 1",
+	))
```
Comment (zz-jason) [inline planner/core/prepare_test.go:709]: please execute another SQL that not found in the plan cache, and check the value of `@@last_statement_found_in_plan_cache` equals to `1`

File: planner/core/prepare_test.go [hunk 8]
```diff
@@ -676,6 +676,39 @@ func (s *testPlanSerialSuite) TestPlanCacheUnionScan(c *C) {
...
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id int)")
+	tk.MustExec("insert into t values (1),(2),(3),(4)")
+	tk.MustExec("prepare stmt from 'select * from t where id=?'")
+	tk.MustExec("set @doma = 1")
+	tk.MustQuery(`show session variables like "last_statement_found_in_plan_cache"`).Check(testkit.Rows(
+		"last_statement_found_in_plan_cache 0",
+	))
```
Comment (zz-jason) [inline planner/core/prepare_test.go:705]:
```suggestion
	tk.MustQuery(`select @@last_statement_found_in_plan_cache`).Check(testkit.Rows("0"))
```

File: planner/core/prepare_test.go [hunk 9]
```diff
@@ -676,6 +676,38 @@ func (s *testPlanSerialSuite) TestPlanCacheUnionScan(c *C) {
...
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id int)")
+	tk.MustExec("insert into t values (1),(2),(3),(4)")
+	tk.MustExec("prepare stmt from 'select * from t where id=?'")
+	tk.MustExec("prepare stmt2 from 'select /*+ ignore_plan_cache() */ * from t where id=?'")
+	tk.MustExec("set @doma = 1")
+	tk.MustQuery(`select @@last_statement_found_in_plan_cache`).Check(testkit.Rows("0"))
+	tk.MustQuery("execute stmt using @doma").Check(testkit.Rows("1"))
+	tk.MustQuery(`select @@last_statement_found_in_plan_cache`).Check(testkit.Rows("1"))
```
Comment (zz-jason) [inline planner/core/prepare_test.go:706]: If we execute a non prepared statements, would `last_statement_found_in_plan_cache` been reset to 0? For example `select 1`;

Comment (lawyerphx) [reply planner/core/prepare_test.go:706]: No, it won't

**Comment by lawyerphx** [comment]
PTAL~ @francis0407

File: sessionctx/variable/session.go [hunk 10]
```diff
@@ -594,6 +594,11 @@ type SessionVars struct {
 	// WindowingUseHighPrecision determines whether to compute window operations without loss of precision.
 	// see https://dev.mysql.com/doc/refman/8.0/en/window-function-optimization.html for more details.
 	WindowingUseHighPrecision bool
+
+	// FoundInPlanCache indicates whether this statement was found in plan cache
```
Comment (francis0407) [inline sessionctx/variable/session.go:598]:
```suggestion
	// FoundInPlanCache indicates whether this statement was found in plan cache.
```

File: sessionctx/variable/session.go [hunk 11]
```diff
@@ -594,6 +594,11 @@ type SessionVars struct {
...
+	FoundInPlanCache bool
+	// PrevFoundInPlanCache indicates whether the last statement was found in plan cache
```
Comment (francis0407) [inline sessionctx/variable/session.go:600]:
```suggestion
	// PrevFoundInPlanCache indicates whether the last statement was found in plan cache.
```

File: sessionctx/variable/session.go [hunk 12]
```diff
@@ -1247,6 +1254,8 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 		atomic.StoreUint64(&config.GetGlobalConfig().Log.QueryLogMaxLen, uint64(tidbOptInt64(val, logutil.DefaultQueryLogMaxLen)))
 	case TiDBCheckMb4ValueInUTF8:
 		config.GetGlobalConfig().CheckMb4ValueInUTF8 = TiDBOptOn(val)
+	case TiDBFoundInPlanCache:
```
Comment (francis0407) [inline sessionctx/variable/session.go:1257]: IMHO, this variable should not be set by user?

Comment (lawyerphx) [reply sessionctx/variable/session.go:1257]: Right. `last_statement_found_in_plan_cache` is a variable that only needs to be read. The set operator used for most system variables seems totally useless here and may cause some puzzles for users.

**Comment by lawyerphx** [comment]
\run-unit-tests

**Comment by eurekaka** [comment]
Can we change the variable name to a simpler one, e.g, `last_plan_from_cache`?

**Comment by lawyerphx** [comment]
> Can we change the variable name to a simpler one, e.g, `last_plan_from_cache`?

I also think the name of this variable is too long, renamed now.

**Pull request merged**